feat: add support for multiple data overlay sources (#12)
MetabolicAtlas/data-files: DATA_OVERLAY.md

## Data source file requirements

- File format: [`tsv`](https://en.wikipedia.org/wiki/Tab-separated_values)
- 1st row: This is the header of the file.
  - 1s column: `id`
  - 2nd-nth columns: each subsequent column should contain the name of the dataset
- 2nd-nth rows: each row should contain values that correspond to the header (1st row)
  - 1st column: component identifier
  - 2nd-nth columns: a decimal value between 0 and 1

### Example data source file contents

| id              | adipose tissue | adrenal gland |
| --------------- | -------------- | ------------- |
| ENSG00000000419 | 0.456          | 0.697         |
| ENSG00000000938 | 0.291          | 0.57          |

### The 0-1 range

When one dataset is selected in the Data overlay, each value in the 0-1 range corresponds to a color in the following spectrum.

<img width="300" alt="image" src="https://user-images.githubusercontent.com/423498/128190295-e70b6b13-f51e-4076-99ee-3094287466d3.png">

The requirement for all values to be in the 0-1 range is to ensure values from different data sources can be handled the same way by the visualization component.

> When adding new data sources, it is important to not excessively limit the range of the values. For example, it might not be good if all of the values are within 0.01 and 0.02, as that would cover a small segment of the color spectrum. Ideally, the minimum value of the data source should be as close to 0 as possible, while the maximum value should be as close to 1 as possible, with as even distribution as possible between for the rest of the values.

Reviewer: I think this concern should be dealt with separately, and either the paragraph should be updated so it sounds more like a statement (eg "in the case of uneven value distribution, this is what is going to happen"), or completely removed.

Author: Can you elaborate on why it should be dealt with separately?

Regarding the evens of distribution, I thought the impact for uneven distribution is implied since the previous sentence mentions that a small total range (0.01 - 0.02) would mean a small segment of the color spectrum is covered. Similarly for uneven distribution, for example if 50% of the data is within 0.01 - 0.02, and 30% is within 0.98 - 0.99, it would have the same problem. Do you think we should be more explicit about this? Also I wanted this to sound more like a recommendation but not a rule, which was the reasoning that this does not sound like a statement.

Reviewer: > Can you elaborate on why it should be dealt with separately?

While the documentation is part of supporting multiple data overlay source in `data-files`, the handling of values can/should be updated (dealt with) separately, as we get a better picture of the frontend implementation.
The way I see it, either we fix the problem so we do not need to inform the user, or we tell them clearly there is a pitfall. In this case, it's easier to just be clearer, for now. 

> Do you think we should be more explicit about this?

Yes! Here's what I had in mind:

```suggestion
When adding new data sources, it is important to inspect the range of the values. For example, if all of the values are within 0.01 and 0.02, only a small segment of the color spectrum will be used. Similarly, for uneven distribution, e.g. if 50% of the data is within 0.01 - 0.02 and 30% is within 0.98 - 0.99, the same visualization problem would be encountered.
Ideally, the minimum value of the data source should be as close to 0 as possible, while the maximum value should be as close to 1 as possible, with as even distribution as possible between for the rest of the values.
```

Author: Ok, that makes sense. I agree it's better to be clearer for now unless we find out that we could simplify the problem.

Thanks for updating the docs, looks great!

Data overlay also allows for two datasets to be selected at the same time. When this is the case, the following spectrum is used instead.

<img width="300" alt="image" src="https://user-images.githubusercontent.com/423498/128190231-8b475022-109e-4905-8a01-5e862c397899.png">

## Folder structure and naming conventions

All data source files should be stored in the [data-files repo](https://github.com/MetabolicAtlas/data-files).

> Under each integrated model, there should be a `dataOverlay` folder, for example: `data-files/integrated-models/Human-GEM/dataOverlay`. If not, it means the model currently does not have any data sources for data overlay and the `dataOverlay` folder should be added.

Reviewer: ```suggestion
Under each folder in [`/integrated-models`](https://github.com/MetabolicAtlas/data-files/tree/main/integrated-models), there should be a `dataOverlay` folder, for example: `data-files/integrated-models/Human-GEM/dataOverlay`. If not, it means the model currently does not have any data sources for data overlay and the `dataOverlay` folder should be added.
```

Author: Great suggestion, thanks!

Inside of the `dataOverlay` folder, there could be one or more folders, where the name of each folder should be the data type. Currently, the types `transcriptomics` and `metabolomics` are supported.

Inside of a data type folder, there could be one or more data source files, such as `hpa.tsv`. There should also be an `index.tsv` file, which contains the metadata for all of the data sources in this folder. When adding a new data source, a new row should be added to the `index.tsv` file with the following values:

- `filename`: name of the new data source file without the extension, for example: `hpa` for `hpa.tsv`
- `name`: name of the data source to be displayed to the end user, for example: `The Human Protein Atlas`
- `link`: website of the data source, for example: `https://www.proteinatlas.org`
- `lastUpdated`: date for when the data source was last updated in `YYYY-MM-DD` format, for example: `2021-08-01`

## Adding a new data source

When incorporating new data sources, it should be done via a pull request in the [data-files repo](https://github.com/MetabolicAtlas/data-files). The above rules and conventions about data format as well as folder structure should be followed.

> If it's difficult to establish how to transform the values into the 0-1 scale, please add a note about this in the pull request description and set the pull request to draft mode.t

Reviewer: There's an extra `t` at the end.

```suggestion
If it's difficult to establish how to transform the values into the 0-1 scale, please add a note about this in the pull request description and set the pull request to draft mode.
```

Inside of the `dataOverlay` folder, there could be one or more folders, where the name of each folder should be the data type. Currently, the types `transcriptomics` and `metabolomics` are supported.

Inside of a data type folder, there could be one or more data source files, such as `hpa.tsv`. There should also be an `index.tsv` file, which contains the metadata for all of the data sources in this folder. When adding a new data source, a new row should be added to the `index.tsv` file with the following values:

- `filename`: name of the new data source file without the extension, for example: `hpa` for `hpa.tsv`
- `name`: name of the data source to be displayed to the end user, for example: `The Human Protein Atlas`
- `link`: website of the data source, for example: `https://www.proteinatlas.org`
- `lastUpdated`: date for when the data source was last updated in `YYYY-MM-DD` format, for example: `2021-08-01`

## Adding a new data source

When incorporating new data sources, it should be done via a pull request in the [data-files repo](https://github.com/MetabolicAtlas/data-files). The above rules and conventions about data format as well as folder structure should be followed.

> If it's difficult to establish how to transform the values into the 0-1 scale, please add a note about this in the pull request description and set the pull request to draft mode.

Reviewer: I think this line is not needed anymore but otherwise this PR looks good to be merged.
@e0 up to you if you would want to tag extra reviewers.

Reviewer: Ah now I see what you mean: `in the pull request description` on line 58 refers to the pull request on line 56. I'd suggest joining the 2 paragraphs for extra clarity.

Somehow I understood that the transformation should be described in the current PR, which did not make much sense to me, as we established that this PR is not about transforming values.

Author: > Ah now I see what you mean: `in the pull request description` on line 58 refers to the pull request on line 56. I'd suggest joining the 2 paragraphs for extra clarity.
> 
> 
> 
> Somehow I understood that the transformation should be described in the current PR, which did not make much sense to me, as we established that this PR is not about transforming values.

Thanks for clarifying. I can see how it can be confusing in the context of this PR. This is in a standalone markdown file and there is a section header for these two paragraphs so I think it is already clear for the intended audience.
